weblate/screenshots/models.py

```diff
@@ -76,3 +86,29 @@ def change_screenshot_assignment(sender, instance, action, **kwargs):
         name="UnusedScreenshot"
     ).exists():
         component_alerts.delay([instance.pk])
+
+
+@receiver(vcs_post_update)
+def update_screenshots_from_repo(sender, component, previous_head: str, **kwargs):
+    screenshots = Screenshot.objects.filter(translation__component=component)
```

Comment:
You can filter screenshots for changed files only directly without need to fetch them all and do additional filtering in Python:

```suggestion
    screenshots = Screenshot.objects.filter(translation__component=component, repository_filename__in=changed_files)
```

PS: This needs a bit of rearranging of the code.